### DISCUSSION TITLE
Why doesn't file upload trigger a Flow?

### DISCUSSION THREAD

**Discussion by erondpowell** [directus/directus#15762] [Archive]
Recently somebody asked me about how to configure a file upload preventer in flows.
For example, if you want to exclude `.svg` files, how would you configure a flow that prevents this file type?

My first thought is:
- `create.item` event hook trigger on `Directus_Files`
- A condition that checks the file's `type` field..... If it is the wrong type:
- A script to throw an error, break the flow, and return an error.

However, at this point, when I set the event hook to detect `directus_files.item.create`, file upload doesn't trigger the flow:

https://user-images.githubusercontent.com/62306630/192343013-71ccfc30-443c-4da5-a123-7c187eb04c40.mov

Since uploading a file creates a corresponding item in `directus_files`, I'm somewhat perplexed that it won't trigger the flow.
Is this is a misconfiguration of the event hook on my end? is this a bug? or does it "not work that way"?

(CC @w0kyj @br41nslug @licitdev)

**Comment by w0kyj** [comment]
To start, there is a specific event trigger scope for "files.upload"

I started to configure one for this use case a few weeks ago, but hit a couple of bugs that have since been fixed. I have not tested again, but basic flow should be:
- Positive Conditional check
    - True -- log if wanted, but otherwise all good and file should upload and associate
    - False
        - May need some testing, worst case, if we can't cancel the upload, then we use a script or delete operation to remove the file

Hope that helps

**Reply by erondpowell** [reply]
In the above in the video I'm just using a log to test the first point of failure: the event hook.

Also it looks like `files.upload` is *only available* for Actions (non-blocking) atm, which do not pause the event.
If we wanted to use a script to throw an error and cancel the flow, we'd need to use **Filter (Blocking)**.

That said, I'm still perplexed... When we upload a file we Directus still creates an item `directus_files`... I'm guessing this event is not *emitted* or detected because its an indirect action that occurs on file upload... but it still feels like it should be possible.

At this point, I'm not so sure blocking the upload is possible (but it really seems like a handy feature @rijkvanzanten)

@w0kyj If you can live with uploading the file, checking its type, then deleting if wrong- we can use an action.
(Although that seems very very sub-optimal, for a number of reasons... 🤔 )
Anyway, if you say that pattern works for you Jon, I think I can do that.

If @licitdev or @br41nslug has a better solution please lmk! 🙏

**Reply by licitdev** [reply]
> Also it looks like files.upload is only available for Actions (non-blocking) atm, which do not pause the event.
If we wanted to use a script to throw an error and cancel the flow, we'd need to use Filter (Blocking).

This is absolutely correct. The `files.upload` filter event is not currently supported but there are plans to support it.

> I'm guessing this event is not emitted

Yes the events are not emitted for file uploads, but other requests to the `/files` endpoint will trigger the events for `directus_files`.

https://github.com/directus/directus/blob/edf050242c3d3b15731fd86bf61dab44328d154d/api/src/services/files.ts#L60

> it really seems like a handy feature

It will be supported in due time. 👍

**Reply by sensedrive** [reply]
> This is absolutely correct. The files.upload filter event is not currently supported but there are plans to support it.
> It will be supported in due time. 

Is there any information about when this will happen? Should we open a ticket here?

**Reply by licitdev** [reply]
> Is there any information about when this will happen?

[Unfortunately, we can not provide estimates on features or fixes](https://docs.directus.io/getting-started/support.html#can-you-give-an-eta-for-this-feature-fix). As of now, the only way to get an accurate timeline or release date is to [sponsor](https://docs.directus.io/getting-started/support.html#sponsored-work) this task. We recognize that this may be an important feature/fix, but we are a small open-source organization with a lot to triage and complete.

> Should we open a ticket here?

Please upvote on https://github.com/directus/directus/discussions/5162 which requires `file.upload` filter event hook to be implemented.

**Reply by sensedrive** [reply]
I would really appreciate this feature and of course I voted for it. Honestly, it also feels more like a bug than a missing feature or improvement. The hook workflow breaks the usual workflow at this point. Also, the collection "directus_files" is selectable in the flows, which is misleading.

To my problem:
Currently, depending on the role of the uploader, we have to set a flag directly on the file entity. Depending on this flag, we want to restrict the visibility of files.
I don't see any solution for this scenario other than using create/update hooks on this entity to set this flag automatically.

**Comment by cdwmhcc** [comment]
I saw this note in the [documentation](https://directus.io/docs/guides/extensions/api-extensions/hooks#filter-events): `System Collection Exceptions
The collections and fields system collections do not emit a read event. The files collection does not emit a create or update event on file upload. The relations collection does not emit a delete event.`
This is also difficult for me to understand. Why can't I prevent certain files that don't meet the requirements from being uploaded?

**Reply by T0chi** [reply]
I'm wondering the same.

**Reply by GuzH09** [reply]
Did someone find a solution for this? I want to avoid file duplicates, but also to avoid someone from spamming 500 uploads per second... But if there's no event, I don't know how this should be tackled

**Reply by flipswitchingmonkey** [reply]
A somewhat dirty workaround is to use `action('files.upload')` and check the user permissions there, then immediately delete the file if they do not match. Unfortunately that does not prevent the upload spamming but at least the files don't linger...

**Reply by ninoalte** [reply]
@licitdev 
its really overdue that there is a filter hook for files, its been 3.5years since this ticket was opened... can we get an update on this?

**Reply by rijkvanzanten** [reply]
@ninoalte this is an archived GitHub discussion, and not something we're keeping up to date or actively monitoring. We manage feature requests through https://roadmap.directus.io/roadmap; posts there go straight to our product team 🙂
